Modernizations (pyproject.toml, package resources, etc)

hseg opened the pull request:
This makes the project more accessible to packagers used to PEP517-formatted repositories. Moreover, build tools are starting to deprecate old configuration mechanisms, see eg [1].

[1]: https://github.com/pypa/pip/issues/6334

(Two things that struck me while testing this out -- we don't appear to be using any `nosetest`-specific features, so I'm surprised we're not just using the ubiquitous `pytest` instead (tests indeed work with it). Also, I'm unsure what `tests/citeproc-test.py` does, and whether it's meant to be executed as a standard test. `pytest` ignores it by default)

Note: One significant change is that `/bin/csl_unsorted` has been made a submodule of `citeproc` (I've created a `citeproc.scripts.*` namespace for these) so as to use standard entry points, which should be more portable across platforms.

Also, this incorporates #167 since I was already working on those bits
Closes: #165, #167

hseg committed:
Migrate to pyproject.toml

This makes the project more accessible to users expecting a
PEP517-formatted repository. Moreover, build tools are starting to
deprecate old configuration mechanisms, see eg [1].

[1]: https://github.com/pypa/pip/issues/6334

tmorrell commented on pyproject.toml:
```diff
@@ -0,0 +1,68 @@
+dynamic = [ "version", "readme" ]
+name = "citeproc-py"
+authors= [ { name = 'Brecht Machiels', email='brecht@mos6581.org' } ]
+urls = { homepage = 'https://github.com/brechtm/citeproc-py' }
```
This should be updated to https://github.com/citeproc-py/citeproc-py

tmorrell reviewed (approved):
This looks like a good modernization. Thanks!

tmorrell commented:
tests/citeproc-test.py gets run separately since it manages what errors are new. https://github.com/citeproc-py/citeproc-py/blob/2aeb25081c5ac66050ef773f81cc64e611ba517b/tox.ini#L19. There might be a more elegant way to handle that, but it seems to work alright.

I don't know why nosetest was used. If pytest works it might be worth switching.

I've tested this locally and it seems to work fine. I've tagged this for a minor release, but will leave the PR open for at least a week in case there is additional feedback.

hseg committed:
Update project URL

hseg committed:
Migrate to pytest

The tests all work unchanged, and pytest's autodetection already picks
up on all of them with no need for configuration.
Also, since tests/citeproc-test.py isn't named test_*.py, it is
automatically ignored.
Preferring pytest over nose given that it's more maintained nowadays.
Added pytest-cov to automatically run coverage for us.

hseg commented:
OK, I updated the URL, fixed some stylistic issues, and moved everything to pytest. Everything should still Just Work, and even better, we need less config.
(Except I forgot to add pytest-cov to tox.ini and am afk, will fix that tomorrow)

hseg commented:
Hm, actually I seem to have broken `tox`, one moment.

hseg commented:
OK, it seems the combination of invoking as `pytest` rather than `python -m pytest` (which didn't add the `citeproc` package to `pythonpath`) and the `changedir` made `pytest` fail to find some files.
I've opted to remove the `changedir`, given that I haven't seen it in any other project (only have seen `{envtmpdir}` used to build the `rply` docs), but another approach is to just append `{toxinidir}` to the `pytest` arguments if that is preferrred.
I've validated these changes by running tox in a venv here, and other than the obvious warnings due to `tests/citeproc-test.py`, everything went fine, so I'm more confident this should work now.

tmorrell reviewed (changes requested):
I'm a bit nervous about the error  OSError: Error reading file '/Users/runner/work/citeproc-py/citeproc-py/citeproc/data/schema/csl.rng': failed to load "/Users/runner/work/citeproc-py/citeproc-py/citeproc/data/schema/csl.rng": No such file or directory`. This should be generated by setup.py. Though there seems to be another missing file error, so maybe it's just something about the current directory. In any event the tests need to pass in order to merge.

hseg commented:
Of course, investigating...

hseg commented:
Oddly enough, when I run the actions on `ubuntu-latest` using https://github.com/nektos/act, I get no errors. But then, perhaps it's doing things differently?

hseg commented:
BTW, the `tests/citeproc-test.py` warnings are very distracting, but I'm unsure how to fix them

hseg commented:
I don't get errors building in a chroot, either

hseg commented:
Wait! Isn't that path part of a submodule?

hseg commented:
Ahah! It appears the custom build hooks are meant to generate that `.rng` file, but for some reason tox isn't invoking them. (That `.rng` file isn't part of the submodule, rather is generated by `convert_rnc` in `setup.py`)

hseg committed:
Rename citeproc.string to prevent shadowing stdlib

hseg commented:
... also, why on earth is `tests/citeproc-test.py` doing its own submodule management?

tmorrell commented:
That's the way it was implemented a decade ago? Way before my time....

The submodules that are part of the code are pretty fixed, while the tests get updated more often. And the approach in the tests is honestly a bit more automated and readable (and avoids confusion like https://github.com/citeproc-py/citeproc-py/issues/172)

hseg commented:
Not sure I agree, but that's an issue for another PR.

Found the proximal cause for the breakage -- it's my dropping of `changedir = {envtmpdir}` from `tox.ini`, but I'd rather figure out how to not need that than revert it.

hseg commented:
Didn't manage to finish this tonight, and I'll be busy over the weekend. Made a little progress in understanding the problem:
IIUC it seems tox is installing the wheel to a venv without running the custom build hooks (possibly it's picking up the wheel from a cache and then the repository gets restored to a pristine state?) The important implication of this is that the rng file does exist inside the installed package, but not in the working directory.
Therefore, I'd recommend we use importlib_resources to look up the csl data files we package, which _should_ work regardless of the state of the working directory (since those lookups will target the package installed in the venv). With that, we should be able to have the code be more robustly location-independent.
The changes to most of the code look relatively simple. However, one corner where I still haven't figured out a way to avoid path computations is in setup.py, though perhaps that isn't so bad (and may be conceptualized as being akin to a Makefile).

tmorrell commented:
I'm supportive of the move to importlib_resources to make pathing more consistent. I'm also fine treating setup.py as a special case. 

Also no rush! Have a good weekend.

hseg committed:
Use package resources over direct reads

This makes the lookups less sensitive to the exact relative locations of
the resources, and additionally unlocks the ability to eg distribute the
project as an executable zip which bundles the resources, see
https://docs.python.org/3/library/zipapp.html

hseg commented:
An update on the situation - I've mostly figured out importlib.resources. However, in changing that, I notice that while CitationStyleLocale can only be passed a locale name, CitationStyleStyle can be passed a filename with and without extension. This makes the logic more complicated, especially since testing for the existence of a file and testing for the existence of a resource use different APIs -- reading is the same, so I can just do something like
((LOCALES_PATH if root is None else root) / f'locale-{locale}.xml').open()
So if I could drop that and just unconditionally use
(root / f'{style}.csl').open()
(and catch errors as they arise), that'd be nice.
Also, I'm planning on replacing the overwriting of STYLES_PATH by adding a root parameter to the CitStySty constructor - is there any appetite for adding one to CitStyLoc as well, for symmetry?

Other than that, I've been trying to simplify setup.py and the dev tooling in general. A couple of ideas I've had which I'd like to run past you:
- instead of vendoring a copy of versioneer, setuptools-scm seems to do all we're using versioneer for, at the cost of slightly different version strings out of the box (it eg adds a date and "dev" if the repo is not at a tagged commit). One advertised feature I haven't explored yet is that it might render MANIFEST.in unnecessary by picking up the files it needs from git's known files
- coverage.py still uses distutils APIs, and in general seems to just be there to provide baked-in defaults. I'd like to explore whether these can't be passed on the command-line, in which case I'd suggest baking them into tox.ini instead
- I'm dissatisfied with the build input generation in setup.py -- particularly the chdir line screams fragility of the process to me, but so far I'm not seeing ways to improve on that, other than orchestrating the build through a Makefile. Perhaps other build backends have such codegen-supporting features, but that might be too high a price to pay for removing a couple lines of code. Needs more exploration
- I still stand by my assertion that tests/citeproc-tests.py shouldn't be doing its own submodule management. Instead, we could have the test-running code in tox.ini first run git submodule update --init to get all submodules to their proper states

(I should also change the PR title to reflect the scope creep 😅, will do so tomorrow)

hseg commented:
Rebased this onto latest master. Will see if I can find a moment soon to review if any work remains to be done here, I do see I had questions back in July that were never addressed.

tmorrell commented:
One thing we added since you started is https://github.com/citeproc-py/citeproc-py/pull/180, which adds more options for the styles. I'm not sure we want to break functionality on what type of style files w/wo extensions are accepted.

I'd be happy to simplify versioneer stuff...but we need to make sure that if works through the whole release and development process. It's just a lot to test.

Overall I'd say smaller change sets are easier to review and get merged. If there is a lot in a single PR it just is more difficult to ensure it's not going to break things.

hseg commented:
17 sept 2025 18:12:23 Tom Morrell ***@***.***>:

> 
>      [Imagen]*tmorrell* left a comment (citeproc-py/citeproc-py#168)[https://github.com/citeproc-py/citeproc-py/pull/168#issuecomment-3303469402]
>     
> 
> One thing we added since you started is #180[https://github.com/citeproc-py/citeproc-py/pull/180], which adds more options for the styles. I'm not sure we want to break functionality on what type of style files w/wo extensions are accepted.
> 
> I'd be happy to simplify versioneer stuff...but we need to make sure that if works through the whole release and development process. It's just a lot to test.
> 
> Overall I'd say smaller change sets are easier to review and get merged. If there is a lot in a single PR it just is more difficult to ensure it's not going to break things.
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/citeproc-py/citeproc-py/pull/168#issuecomment-3303469402], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AAJTW5JA2O3K6WPGKZTDE633TF25NAVCNFSM6AAAAACAO76FYCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGMBTGQ3DSNBQGI].
> You are receiving this because you authored the thread.
> [Imagen de rastreo][https://github.com/notifications/beacon/AAJTW5OBADCEDNLGE2WSBLL3TF25NA5CNFSM6AAAAACAO76FYCWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTWE43YVU.gif]
> 

Fair enough, I think I'll split this PR up into a couple smaller, more focused ones (though iirc I made an effort to make the _commits_ small and self-contained)

yarikoptic commented:
@hseg thanks in advance. I will move it to draft for now, move out when it is ready for re-review.

yarikoptic commented:
@hseg ping on this PR

hseg commented:
Ah, right - it had fallen off my agenda, I'll see if I can find an evening this week to work on this.

hseg commented:
OK, part 1 (`pyproject.toml`) done, I'll do the remaining parts another time.

Remaining cannibalizable bits of this PR:
- [ ] Use `importlib.resources` instead of `package_data`
- [ ] Convert `nosetests` usage to `pytest`
- [ ] Avoid duplicating git submodule management in `tests/citeproc-latest.py` (perhaps by having `tox` make sure `git` populates those submodules first?)
